### DISCUSSION TITLE
Execution order of migrations

### DISCUSSION THREAD

**Discussion by andypimlett** [adonisjs/core#2345] [Help] [answered]
Do migrations have to be created in such a way that tables are created before foreign key constraints? If I attempt to run a migration that has a relationship against a table that hasn't yet run its migration its fails based on no relationship existence. Seems odd that the migrations would have a dependancy that requires the database be constructed in an exact order.

Appreciate that would be the case when creating via SQL but I assumed the power of migrations was to allow the logic to automate the implementation of the schema definitions?

This being the case is it best to create an initial set of migrations to create tables with no relationships and then write all of the relationships into a set of migration extensions?

**Comment by McSneaky** [comment, marked as answer]
Answered in Discord, but will give answer in here too in case someone lands in here thro search engines

Framework is not analysing your migrations in any way, it just runs them in alphabetical order. You can reorder your migrations by renaming them to something like `0_first_stuff.ts`, `1_more_things.ts`, `2_stuff.ts` etc

> This being the case is it best to create an initial set of migrations to create tables with no relationships and then write all of the relationships into a set of migration extensions?

That depends on your DB engine and limitations, but in case you have 2 or 3 tables (3rd as pivot table). Then you can first setup table 1, then table 2 and define relations in table 3. Or without pivot table create first table as-is and then in other table you can set relationships to first table
Some DBs also allow to disable all checks while creating migrations, then you can migrate and add relations etc in random order, but that's real bad practise and something might end up beeing really broken like that
